oidc_middleware.go

```diff
@@ -0,0 +1,163 @@
+	for _, key := range body.Keys {
+		kid := key.KeyID
+		rsakey := new(rsa.PublicKey)
+		number, err := base64.RawURLEncoding.DecodeString(key.Number)
+		if err != nil {
+			return nil, fmt.Errorf("decode JWT 'n' field: %w", err)
+		}
+		rsakey.N = new(big.Int).SetBytes(number)
+		rsakey.E = rsaExponent
+		rsaKeys[kid] = rsakey
+	}
+	return rsaKeys, nil
+}
+
+func newOIDCMiddleware(rsaKeys map[string]*rsa.PublicKey, config OIDCConfig) *oidcMiddleware {
+	return &oidcMiddleware{
+		rsaKeys: rsaKeys,
+		config:  config,
+	}
+}
+
+type oidcMiddleware struct {
+	rsaKeys map[string]*rsa.PublicKey
+	config  OIDCConfig
+}
+
+// VerifyTokenMiddleware is a gin middleware function that enforces validity of the access bearer token on every
+// request. This uses the environment vars WHARF_HTTP_OIDC_IssuerURL and WHARF_HTTP_OIDC_AudienceURL as limiters
+// that control the variety of tokens that pass validation.
+func (m *oidcMiddleware) VerifyTokenMiddleware(ginContext *gin.Context) {
+	if m.rsaKeys == nil {
+		ginutil.WriteProblem(ginContext, problem.Response{
+			Type:   "/prob/api/oidc/missing-rsa-keys",
+			Title:  "Missing OIDC public keys.",
+			Status: http.StatusInternalServerError,
+			Detail: "The OIDC RSA public keys were not properly set up during initialization of the wharf-api.",
+		})
+		ginContext.Abort()
+		return
+	}
+	isValid := false
+	errorMessage := ""
+	tokenString := ginContext.Request.Header.Get("Authorization")
+	if !strings.HasPrefix(tokenString, "Bearer ") {
+		ginutil.WriteUnauthorized(ginContext, "Expected authorization scheme to be 'Bearer' (case sensitive), but was not.")
+		ginContext.Abort()
+		return
+	}
+	tokenString = strings.TrimPrefix(tokenString, "Bearer ")
+	token, err := jwt.Parse(tokenString, func(token *jwt.Token) (interface{}, error) {
+		if kid, ok := token.Header["kid"].(string); ok {
+			return m.rsaKeys[kid], nil
+		}
+		return nil, errors.New("expected JWT to have string 'kid' field")
+	})
+	if err != nil {
+		errorMessage = err.Error()
+	} else if !token.Valid {
+		errorMessage = "invalid access bearer token."
+	} else if token.Header["alg"] == nil {
+		errorMessage = "missing 'alg' field."
+	} else if token.Claims.(jwt.MapClaims)["aud"] != m.config.AudienceURL {
+		errorMessage = "invalid 'aud' field."
+	} else if iss, ok := token.Claims.(jwt.MapClaims)["iss"].(string); !ok {
+		errorMessage = "invalid or missing 'iss' field: should be string."
+	} else if !strings.Contains(iss, m.config.IssuerURL) {
+		errorMessage = "invalid 'iss' field: disallowed issuer."
+	} else {
+		isValid = true
+	}
+	if !isValid {
+		ginutil.WriteUnauthorized(ginContext, "Invalid JWT: "+errorMessage)
+		ginContext.Abort()
+	}
+}
+
+// SubscribeToKeyURLUpdates ensures new keys are fetched as necessary.
+// As a standard OIDC login provider keys should be checked for updates ever 1 day 1 hour.
+func (m *oidcMiddleware) SubscribeToKeyURLUpdates() {
+	fetchOidcKeysTicker := time.NewTicker(m.config.UpdateInterval)
```

Comment:
convention
```suggestion
	fetchOIDCKeysTicker := time.NewTicker(m.config.UpdateInterval)
```

Reply:
We eariler spent a long time deciding on the non-capitalised version of this previously in the PR so ill skip this change.